### PR TITLE
[core] Introduce consume-id for retore consuming and reserve snapshot

### LINKED ISSUE #881: [Feature] Support consume group id like kafka
### Search before asking

- [X] I searched in the [issues](https://github.com/apache/incubator-paimon/issues) and found nothing similar.

### Motivation

Consume group id in kafka is useful for restarting without state.
We can use files to record consume group id too.

### Solution

_No response_

### Anything else?

_No response_

### Are you willing to submit a PR?

- [ ] I'm willing to submit a PR!

### PR DESCRIPTION
**Pull request by JingsongLi**
fix #881

### Purpose

This PR is intended to introduce a consume mode, when a `consume-id` is specified, scan will writes the progress of the consume to the file system, which has several advantages:
1. When the job is killed, the newly started job can continue to consume from the previous progress without resuming from the state.
2. When deleting expired snapshots, it ensures that the snapshot being consumed will not be deleted, so that the consumer will not read the expired snapshot and get hung.

This PR also introduces a new Source implementation:
- MonitorFunction
- ReadOperator

This is to avoid Flink's Source interface forcing a checkpoint on Splits, which may only read part of the snapshot data when consuming, the progress of consuming cannot be recorded.

### Documentation

Added.

### CODE REVIEW

File: paimon-core/src/main/java/org/apache/paimon/table/source/AbstractInnerTableScan.java [hunk 1]
```diff
@@ -65,6 +69,16 @@ protected StartingScanner createStartingScanner(boolean isStreaming) {
             return new ContinuousCompactorStartingScanner();
         }
 
+        // read from consumer id
+        String consumerId = options.consumerId();
```
Comment (lilyzhoupeijie) [inline paimon-core/src/main/java/org/apache/paimon/table/source/AbstractInnerTableScan.java:73]: do we need to compare the snapshot from consumer with snapshot from state?

File: paimon-core/src/main/java/org/apache/paimon/table/source/InnerStreamTableScanImpl.java [hunk 2]
```diff
@@ -194,4 +195,18 @@ public Long checkpoint() {
...
+    public void notifyCheckpointComplete(@Nullable Long nextSnapshot) {
+        if (nextSnapshot == null) {
+            return;
+        }
+
+        String consumerId = options.consumerId();
+        if (consumerId != null) {
+            snapshotSplitReader
+                    .consumerManager()
+                    .recordConsumer(consumerId, new Consumer(nextSnapshot));
```
Comment (lilyzhoupeijie) [inline paimon-core/src/main/java/org/apache/paimon/table/source/InnerStreamTableScanImpl.java:209]: If two flink programs with the same consumer-id exist simultaneously, what semantics do we guarantee? or  it is necessary we try to  guarantee? cause there is no lock on the file,and anyone with the consumer-id can change it.

File: paimon-core/src/main/java/org/apache/paimon/table/source/AbstractInnerTableScan.java (hunk 1 above)
Comment (JingsongLi) [reply paimon-core/src/main/java/org/apache/paimon/table/source/AbstractInnerTableScan.java:73]: I think if the relevant content is recorded inside the state, it should not be read from the consumer file and should be trusted in the state.
Do you think something bad will happen?

File: paimon-core/src/main/java/org/apache/paimon/table/source/InnerStreamTableScanImpl.java (hunk 2 above)
Comment (JingsongLi) [reply paimon-core/src/main/java/org/apache/paimon/table/source/InnerStreamTableScanImpl.java:209]: Good point!
I'm not sure what kind of behavior it should be, so I'll let them each play it out here, not sure who will record success.

**Comment by JingsongLi** [comment]
Thanks @lilyzhoupeijie for the review, do you have any thoughts on the solution? Does the current solution work to your satisfaction?

**Comment by lilyzhoupeijie** [comment]
maybe each consumer has an inner UUID store in state?and use UUID +consumer-id to update the record file and then we can decide which semantics  to use.
or compare the snapshot in record file with current snapshot -1,and then to decide whether to update? if not equals,it means that some other consumer has finished reading the snapshot.something like java CAS to avoid ABA ?
or use the feature in storage,such as hdfs,it provide lease to keep single writer,keep only one consumer writing the record file

**Comment by JingsongLi** [comment]
> maybe each consumer has an inner UUID store in state?and use UUID +consumer-id to update the record file and then we can decide which semantics to use. or compare the snapshot in record file with current snapshot -1,and then to decide whether to update? if not equals,it means that some other consumer has finished reading the snapshot.something like java CAS to avoid ABA ? or use the feature in storage,such as hdfs,it provide lease to keep single writer,keep only one consumer writing the record file

The discussion is based on the premise that we need to support a third requirement:

3. the same consumer-id supports multiple simultaneous reads, and the same snapshot needs to be guaranteed to be read by only one reader. It should behave just like the groupId of the message queue.

I think it's an interesting topic, but I'm not sure if the current users have such a need, and I propose that I can keep it for now.
The current information is enough, for example I know which snapshot I read and I know how I need to update the consumer file, similar to CAS (compare and swap), but the hard part is the subsequent processing, really to avoid it, I guess we can only have pre-commit and commit phases.

**Comment by lilyzhoupeijie** [comment]
> > maybe each consumer has an inner UUID store in state?and use UUID +consumer-id to update the record file and then we can decide which semantics to use. or compare the snapshot in record file with current snapshot -1,and then to decide whether to update? if not equals,it means that some other consumer has finished reading the snapshot.something like java CAS to avoid ABA ? or use the feature in storage,such as hdfs,it provide lease to keep single writer,keep only one consumer writing the record file
> 
> The discussion is based on the premise that we need to support a third requirement:
> 
> 3. the same consumer-id supports multiple simultaneous reads, and the same snapshot needs to be guaranteed to be read by only one reader. It should behave just like the groupId of the message queue.
> 
> I think it's an interesting topic, but I'm not sure if the current users have such a need, and I propose that I can keep it for now. The current information is enough, for example I know which snapshot I read and I know how I need to update the consumer file, similar to CAS (compare and swap), but the hard part is the subsequent processing, really to avoid it, I guess we can only have pre-commit and commit phases.

Unless we have a global lock, maybe we cannot prevent data from being consumed multiple times.But is there some way to avoid sending overconsumed data to downstream operators?so that we still can keep an end-to-end eos.
similar with:
https://github.com/apache/incubator-paimon/issues/881
is this still needed?maybe global lock or pre-allocation based on snapshot.if not needed,this one can be closed.

File: paimon-flink/paimon-flink-common/src/main/java/org/apache/paimon/flink/source/operator/MonitorFunction.java [hunk 3]
```diff
@@ -0,0 +1,222 @@
...
+            }
+
+            // given that the parallelism of the function is 1, we can only have 1 retrieved items.
+
+            Preconditions.checkArgument(
+                    retrievedStates.size() <= 1,
+                    getClass().getSimpleName() + " retrieved invalid state.");
+
+            if (retrievedStates.size() == 1) {
+                this.scan.restore(retrievedStates.get(0));
```
Comment (FangYongs) [inline paimon-flink/paimon-flink-common/src/main/java/org/apache/paimon/flink/source/operator/MonitorFunction.java:130]: Should we need to restore scan with the snapshot id in state or in consume-id file here?

**Comment by JingsongLi** [comment]
> > > maybe each consumer has an inner UUID store in state?and use UUID +consumer-id to update the record file and then we can decide which semantics to use. or compare the snapshot in record file with current snapshot -1,and then to decide whether to update? if not equals,it means that some other consumer has finished reading the snapshot.something like java CAS to avoid ABA ? or use the feature in storage,such as hdfs,it provide lease to keep single writer,keep only one consumer writing the record file
> > 
> > 
> > The discussion is based on the premise that we need to support a third requirement:
> > 
> > 3. the same consumer-id supports multiple simultaneous reads, and the same snapshot needs to be guaranteed to be read by only one reader. It should behave just like the groupId of the message queue.
> > 
> > I think it's an interesting topic, but I'm not sure if the current users have such a need, and I propose that I can keep it for now. The current information is enough, for example I know which snapshot I read and I know how I need to update the consumer file, similar to CAS (compare and swap), but the hard part is the subsequent processing, really to avoid it, I guess we can only have pre-commit and commit phases.
> 
> Unless we have a global lock, maybe we cannot prevent data from being consumed multiple times.But is there some way to avoid sending overconsumed data to downstream operators?so that we still can keep an end-to-end eos. similar with: #881 is this still needed?maybe global lock or pre-allocation based on snapshot.if not needed,this one can be closed.

We can disregard it for now. In fact, a simple case oriented to multiple reads is to report an error.

**Comment by JingsongLi** [comment]
Thanks @lilyzhoupeijie @yuzelin for your review, merged...

**Pull request merged**